### PR TITLE
Ptime

### PR DESCRIPTION
**Pull request by dinosaure**
I add a conditionnal compilation in `syndic` to compile with `calendar` by default or with `ptime` if it is available. I produce a good `META` and avoid the unix dependency with `ptime`.

### CODE REVIEW

**Comment by dinosaure** [comment]
May be it's time to update the CI (like [Decompress](https://github.com/oklm-wsh/Decompress/blob/master/.travis.yml)).

**Comment by dinosaure** [comment]
Travis is happy :+1: !

Travis happy

**Comment by hannesm** [comment]
let me ask you a stupid question: why present the user a choice between Calendar and Ptime?  IMHO one of the two is sufficient (and since Calendar depends on Unix, I'd clearly prefer Ptime)

**Comment by dinosaure** [comment]
I wait the position of @Chris00 to switch to Calendar or Ptime :) .

File: _oasis [hunk 1]
```diff
@@ -39,9 +48,9 @@ Executable main
   BuildDepends:         syndic
 
 SourceRepository master
-  Type:       git
-  Location:   https://github.com/Cumulus/Syndic.git
-  Browser:    https://github.com/Cumulus/Syndic
+  Type:                 git
+  Location:             https://github.com/Cumulus/Syndic.git
+  Browser:              https://github.com/Cumulus/Syndic
```
Comment (Chris00) [inline _oasis]: Point?

Comment (dinosaure) [inline _oasis]: It's not the subject of the PR but I don't like which is not aligner :$

File: lib/ptime/syndic_date.ml [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+let month_of_date =
+  let months = [| Jan; Feb; Mar; Apr; May; Jun; Jul; Aug; Sep; Oct; Nov; Dec |] in
+  fun t ->
+  let (_, i, _) = Ptime.to_date t in
+  months.(i)
+
+(* RFC3339 date *)
+let of_rfc3339 s = Ptime.of_rfc3339 s
+  |> function Result.Ok (t, Some tz_offset_s, _) ->
```
Comment (Chris00) [inline lib/ptime/syndic_date.ml]: Why not `match Ptime.of_rfc3339 s with`?

File: lib/ptime/syndic_date.ml [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+              Ptime.to_date_time ~tz_offset_s t
+              |> Ptime.of_date_time
+              |> (function Some x -> x
```
Comment (Chris00) [inline lib/ptime/syndic_date.ml]: Same, why not `(match Ptime.of_date_time with ...)`?

File: lib/ptime/syndic_date.ml [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+                         | None -> invalid_arg(sprintf "Syndic.Date.of_string: cannot parse %S" s))
+            | Result.Ok (t, None, _) ->
+              Ptime.to_date_time t |> Ptime.of_date_time
+              |> (function Some x -> x
+                         | None -> invalid_arg(sprintf "Syndic.Date.of_string: cannot parse %S" s))
+            | _ -> invalid_arg(sprintf "Syndic.Date.of_string: cannot parse %S" s)
```
Comment (Chris00) [inline lib/ptime/syndic_date.ml]: Do we really want to use a catch-all?

File: lib/ptime/syndic_date.ml [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+let to_rfc3339 d =
+  (* Example: 2014-03-19T15:51:25.050-07:00 *)
+  Ptime.to_rfc3339 d
+
+(* Convenience functions *)
+
+let day_of_week =
+  let wday = [| `Thu; `Fri; `Sat; `Sun; `Mon; `Tue; `Wed |] in
```
Comment (Chris00) [inline lib/ptime/syndic_date.ml]: It is odd to use polymorphic variants here if standard variants are used for months.

File: lib/ptime/syndic_date.ml [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    let i = fst Ptime.(Span.to_d_ps @@ to_span t) mod 7 in
+    wday.(if i < 0 then 7 + i else i)
+
+let string_of_day = function
+  | `Thu -> "Thu" | `Fri -> "Fri" | `Sat -> "Sat" | `Sun -> "Sun"
+  | `Mon -> "Mon" | `Tue -> "Tue" | `Wed -> "Wed"
+
+let year t =
+  Ptime.to_date t
+  |> fun (year, _, _) -> year
```
Comment (Chris00) [inline lib/ptime/syndic_date.ml]: Why not a `let (year,_,_) = ...`?

**Comment by Chris00** [comment]
I had a look and I think we can unconditionally move to Ptime.  ocaml.org code will need to be updated but that's OK.  One should also let the author of `river` know because [it relies on Syndic](http://opam.ocaml.org/packages/syndic/syndic.1.3.1/).

**Comment by dinosaure** [comment]
Thanks for your feedback, I will improve the code with your comments :)

**Pull request merged**